File: StructuralPattern/Flyweight/Python/Factory/GameObjectFactory.py

from Interfaces import GameObject
from Concretes import Plant, Zombie

class GameObjectFactory:
    """
        Factory for creating GameObjects
    """
    def __init__(self, *args, **kwargs):
        self.gameObjectPool = {} # HashMap: Map <String, GameObject>
        pass
    
    def get_Plant(self, type, image) -> GameObject:
        key = "Plant-" + type
        if key not in self.gameObjectPool:
            self.gameObjectPool[key] = Plant(type, image)
        return self.gameObjectPool[key]
    
    def get_Zombie(self, type, image) -> GameObject:
        key = "Zombie-" + type
        if key not in self.gameObjectPool:
            self.gameObjectPool[key] = Zombie(type, image)
        return self.gameObjectPool[key]

File: StructuralPattern/Flyweight/Golang/Factory/GameObjectFactory.go

package Factory

import (
	"v/Interfaces"
	"v/Concretes"
)

// Flyweight Factory
type GameObjectFactory struct {
    gameObjectPool map[string](iface.GameObject)
}

// Construct a GameObjectFactory
func NewGameObjectFactory() *(GameObjectFactory) {
	return &GameObjectFactory{gameObjectPool: make(map[string](iface.GameObject))}
}

func (gf GameObjectFactory) GetPlant(t string, image string) iface.GameObject {
    key := "Plant-" + t
    check := gf.gameObjectPool[key]
    if check == nil {
        gf.gameObjectPool[key] = Concretes.NewPlant(t, image)
    }
    return gf.gameObjectPool[key]
}

func (gf GameObjectFactory) GetZombie(t string, image string) iface.GameObject {
    key := "Zombie-" + t
    check := gf.gameObjectPool[key]
    if check == nil {
        gf.gameObjectPool[key] = Concretes.NewZombie(t, image)
    }
    return gf.gameObjectPool[key]
}


